HP85: support for I/O slots & HP-IB interface module

fulivi opened the pull request:
Hi.
This PR introduces the handling of I/O slots in HP85 emulation.
It also introduces the emulation of HP82937 module that interfaces HP85 with HPIB peripherals (with HP9895 disk drive, for example).
By using the Mass storage ROM it's now possible to read & write floppy disks in HP9895 drive.
I'm going to e-mail the dump of 82937 firmware to the usual "code" address.
Thanks.
--F.Ulivi

cuavas commented on src/devices/bus/hp80_io/82937.cpp:
```diff
@@ -0,0 +1,354 @@
+
+// Bit manipulation
+namespace {
+	static constexpr unsigned BIT_MASK(unsigned n)
```
There's no need to use `static` here:
* `constexpr` implies `inline`, which suppresses symbol export already
* It's in an anonymous namespace which suppresses symbol export

cuavas commented on src/devices/bus/hp80_io/82937.cpp:
```diff
@@ -0,0 +1,354 @@
+		return 1U << n;
+	}
+
+	template<typename T> void BIT_SET(T& w , unsigned n)
```
This will behave strangely if the size of `T` is larger than `unsigned` due to `BIT_MASK` always returning unsigned.

cuavas commented on src/devices/bus/hp80_io/82937.cpp:
```diff
@@ -0,0 +1,354 @@
+static constexpr unsigned P1_EOI_BIT = 3;
+static constexpr unsigned P1_DAV_BIT = 2;
+static constexpr unsigned P1_NDAC_BIT = 1;
+static constexpr unsigned P1_NRFD_BIT = 0;
```
Qualifying these these `static` adds clutter:
* `constexpr` integers at global scope imply static (ODR only applies at compilation unit level)
* You could just put them in the anonymous namespace anyway to save the clutter of the keyword on every line.

cuavas commented on src/devices/bus/hp80_io/82937.cpp:
```diff
@@ -0,0 +1,354 @@
+	if (m_updating) {
+		return;
+	}
+	m_updating = true;
```
This pattern for avoiding mutual recursion is usually a symptom of a less-than-ideal design.  Is there a way to rethink the bus to avoid the need to use this pattern?

(Also, for this to be exception-safe, you need to use a local object destructor or a catch-all block to ensure `m_updating` will be cleared.  Doing it this way can lead to a stuck flag.)

cuavas commented on src/devices/bus/hp80_io/82937.h:
```diff
@@ -0,0 +1,71 @@
+	DECLARE_WRITE8_MEMBER(dio_w);
+	DECLARE_READ8_MEMBER(switch_r);
+	DECLARE_WRITE8_MEMBER(latch_w);
+	DECLARE_WRITE_LINE_MEMBER(ieee488_ctrl_w);
```
Read/write members used from your own `MACHINE_CONFIG_MEMBER` functions shouldn't really be public if the signals aren't exposed outside the device itself (stuff used in address maps still needs to be public because of how address maps are implemented).

cuavas commented on src/devices/machine/1mb5.cpp:
```diff
@@ -0,0 +1,295 @@
+
+// Bit manipulation
+namespace {
+	static constexpr unsigned BIT_MASK(unsigned n)
```
Same comment about static as with the other file

cuavas commented on src/devices/machine/1mb5.cpp:
```diff
@@ -0,0 +1,295 @@
+		return 1U << n;
+	}
+
+	template<typename T> void BIT_CLR(T& w , unsigned n)
```
Same comment about what will happen if these are used with `T` larger than `unsigned`

fulivi commented:
Ok, I suppose I'll be addressing these issues in a future patch because the PR has now been merged (or hasn't it?).
The two main points here as I see them:
* "unsigned" values in BIT* templates: the only way to be sure never to overflow would be to use 64 bit integers, which, in my opinion, are a bit overkill here (where I know for sure that I'm dealing with 16 bits at most). Is there a way to parametrize templates on needed size?
* The anti-recursion flag: this design is a consequence of IEEE488 bus design. When you are setting output signals, the changes are immediately reported through CBs. I could probably temporarily disable CBs (by using the flag and making it exception-proof). I'm open to suggestions on improving this. To me too it doesn't look like super-polished design..

cuavas commented:
The stuff with `static` interacting with `constexpr` and the types are not critical, you can clean it up next time you have something else to submit.  No need for a PR just for that.

You can do something like:
```
template <typename T> constexpr unsigned BIT_MASK(unsigned n) { return T(1) << n; }
```

Then use it like:
```
template <typename T> void BIT_SET(T& w , unsigned n)
{
    w |= BIT_MASK<T>(n);
}
```

Not sure what to do about the IEEE488 stuff, it's possibly a limitation with the design of that stuff in MAME.
